app/services/amr/upsert_validated_readings_for_a_meter.rb

```diff
@@ -0,0 +1,52 @@
+module Amr
+  class UpsertValidatedReadingsForAMeter
+    NAN_READINGS = Array.new(48, Float::NAN).freeze
+
+    def initialize(dashboard_meter)
+      @dashboard_meter = dashboard_meter
+    end
+
+    def perform
+      Rails.logger.info "Processing: #{@dashboard_meter} with mpan_mprn: #{@dashboard_meter.mpan_mprn} id: #{@dashboard_meter.external_meter_id}"
+      amr_data = @dashboard_meter.amr_data
+
+      Upsert.batch(AmrValidatedReading.connection, AmrValidatedReading.table_name) do |upsert|
+        amr_data.values.each do |one_day_read|
+          upsert_from_one_day_reading(@dashboard_meter.external_meter_id, upsert, one_day_read)
+        end
+      end
+
+      dates_to_delete = existing_dates_to_be_deleted(amr_data.keys)
+      deleted_count = delete_by_dates(dates_to_delete)
+      Rails.logger.info "Processing: #{@dashboard_meter} - deleted: #{deleted_count}"
+    end
+
+  private
+
+    def upsert_from_one_day_reading(meter_id, upsert, one_day_reading)
+      return if is_nan?(one_day_reading)
+      upsert.row({ meter_id: meter_id, reading_date: one_day_reading.date },
+        meter_id: meter_id,
+        reading_date: one_day_reading.date,
+        kwh_data_x48: one_day_reading.kwh_data_x48,
+        one_day_kwh: one_day_reading.one_day_kwh,
+        substitute_date: one_day_reading.substitute_date,
+        status: one_day_reading.type,
+        upload_datetime: one_day_reading.upload_datetime
+      )
+    end
+
+    def existing_dates_to_be_deleted(new_validated_dates)
```

Comment:
I think this functionality around deleting readings post upserting *should probably* be in a separate service, run after the upsert.

Reply:
yes - I wasn't sure about the size of where clauses - i.e. is there a data limit to what is allowed?